### DISCUSSION TITLE
Templating values in ODK form to avoid duplication of records

### DISCUSSION THREAD

**Discussion by Nik720** [Samagra-Development/SamagraX-community#31] [Installation and Configuration]
@aayushkedia95 Please find the below excel sheet with our existing use cases. Please help us to how can we set a template and use it dynamically to generate links and questions as we discussed in the call.

https://drive.google.com/file/d/1dKjmNx596QKzMAOgde67DDhMnzmr1eBJ/view?usp=sharing

cc - @vinukumar-vs @vrayulu

**Comment by Nik720** [comment]
@aayushkedia95 any update on this?

**Reply by aayushkedia95** [reply]
@Nik720, we will unblock ODK next week for the team. Apologies for the delay.

**Comment by aayushkedia95** [comment]
@vinukumar-vs @Nik720, can we get on a call on Thursday/Friday to unblock the entire ODK Global Flow Component?

Would be great if you can share all questions/doubts by tomorrow EoD. We will try to make required changes to the code before our call. 

cc - @ChakshuGautam @Krishnaj20

**Reply by aayushkedia95** [reply]
@vinukumar-vs @Nik720, please can you share all the ODK related questions here?

cc - @ChakshuGautam @Krishnaj20

**Reply by vinukumar-vs** [reply]
- How can we use a single ODK form for both WhatsApp & Portal with minor text/link changes for some workflow or options.
- If we use a single ODK form how can we handle the user input from WhatsApp(numbers) & from portal/Tara bot(Menu/Option clicks)
- How can we dynamically replace the response string with a user-selected value or reference key. So that we can configure the same URL for all the options(especially for states) & replace only state name dynamically based on user-selected.
- As we discussed in our previous calls, how can we replace number with key-based(choices sheet "name" column). So that any new data addition doesn't require the changes of existing values.

**Reply by aayushkedia95** [reply]
Thank you @vinukumar-vs!

**Reply by aayushkedia95** [reply]
Hi @vinukumar-vs, 

Please refer to the following [documentation](https://www.notion.so/Guides-3965a5286e4f421390024035704ab4e7)

It has two sections - 
1. Building an Adapter
2. ODK for Global menu

We will clarify any further doubts on call. 

cc - @Krishnaj20 @ChakshuGautam

**Reply by Nik720** [reply]
Hi @aayushkedia95 

Can you please provide us access to documentation

**Comment by vinukumar-vs** [comment]
Thanks for the update. We will go through it & add comments if required.

+Nikunj Beladiya ***@***.***>


Regards,
Vinu Kumar V S
Lead Software Engineer
Tel.: +91  <+917259621000>9964300623
Skype: vinukumar_vs
Tarento Technologies Pvt Ltd
www.tarento.com



Please note that this message may contain privileged and confidential
information. This communication and any attachments are solely intended for
the person(s) or organization(s) to whom it is addressed. If you are not
the intended recipient, you are hereby notified that any unauthorized use,
copying or disclosure of this communication and its attachments is strictly
prohibited and may be unlawful. If you have received this communication
without being the intended recipient please discard the email without any
usage and/or distribution of the information and inform us.

**Reply by aayushkedia95** [reply]
Have given access @Nik720! Please check.

**Reply by Nik720** [reply]
@aayushkedia95  As specified in [Documentation](https://www.notion.so/How-tos-ODK-for-Global-Menu-2c063ffca4a74a2f9b1bb21a8e3525c7) will use a hidden variable for channel & provider to make it a common form for portal and WhatsApp bot. 
- how will know the channel & provider value? Do we have to specify it in form?
- Can you please give a sample example of how can we use the **if-else** condition in the label column? I have tried with below but it didn't work. it is showing the same statement as a text in ODK collect app.
 `if(${state_course_board_name} = 'other') OTHER LINK else BOARD LINK`

cc @vinukumar-vs @Krishnaj20

**Reply by aayushkedia95** [reply]
@Nik720, please find my answers below.

1. It can be dynamically pre-filled since you would know the channel and provider using the Adapter API Gateway endpoint - /channel/provider 
2. Please check rows 20, 21 from this sample - https://docs.google.com/spreadsheets/d/1d_u-2cd-8uJOR-x5ork1IWZ4oynCxE5vGQcqivTMUhM/edit#gid=206583462

**Reply by Nik720** [reply]
@aayushkedia95 I have tried the condition as per your example but it is not working when I generate a dynamic URL.  Below is a condition that I have added in the calculation column when I am trying to build a dynamic URL using `${state_course_board_name} & ${calculated_utm_params}` hidden variable but it is not giving valid data. 
I have tested it on ODK collect app. Can you please let me know how to generate this type of data?

`if(${state_course_board_name} = 'other',  'https://diksha.gov.in/explore/1?primaryCategory=course&selectedTab=all&appliedFilters=true&utm_source=${calculated_utm_params}' , 'https://diksha.gov.in/explore/1?primaryCategory=course&selectedTab=all&se_boards=state%20${state_course_board_name}&appliedFilters=true&utm_source=${calculated_utm_params}')`

ODK Form URL: https://docs.google.com/spreadsheets/d/1jhlHrRwSdI3mdo8IjoRExIGHuMxj1Wgg2alXYloi-No/edit?usp=sharing
Form id: sunbird_tara_bot_form_v5
 
cc @vinukumar-vs @Krishnaj20

**Reply by aayushkedia95** [reply]
Sure, @Nik720. Let us check and get back in sometime.

**Reply by aayushkedia95** [reply]
Hey @Nik720, we have corrected the form. Please find it below.  

[Tara-bot.xlsx](https://github.com/Samagra-Development/SamagraX-community/files/6669201/Tara-bot.xlsx)

The ODK syntax was incorrect, in the form you shared. Refer to the following [link](https://docs.getodk.org/) for ODK syntax.

**Reply by Nik720** [reply]
Thanks @aayushkedia95

**Reply by Nik720** [reply]
> @Nik720, please find my answers below.
> 
> 1. It can be dynamically pre-filled since you would know the channel and provider using the Adapter API Gateway endpoint - /channel/provider

@aayushkedia95  I am not getting channel value. I have used channel and provider value as 
` "channel": "WhatsApp",  "provider": "gupshup",` and in ODK form I have added condition based on the channel but I think channel value is not being pre-filled with this value. 

https://docs.google.com/spreadsheets/d/1jhlHrRwSdI3mdo8IjoRExIGHuMxj1Wgg2alXYloi-No/edit#gid=206583462&range=17:17 here I have printed channel value but I am not getting anything on the WhatsApp bot. 
Can you please help with this

form id: sunbird_tara_v5
bot id: e124de54-3020-4972-bb18-f8f81ba5631d
adapter: e124de54-3020-4972-bb18-f8f81ba5631d
      
cc @vinukumar-vs @Krishnaj20

**Reply by aayushkedia95** [reply]
This will be added in runtime. We need to push an update to add this. Will keep you informed @Nik720.

**Reply by Nik720** [reply]
@aayushkedia95 When will we expect to get the update?

**Reply by Nik720** [reply]
@aayushkedia95 Here is a couple of doubts, can you please look into this 

1) with the key-based approach I have to type the same option to get a response and it is working fine, but it is not working with the numeric key as you specified in the documentation that we can still get a response with 1, 2 option
2) still have to type (.) after numeric key to get a response
3) can you please provide the netcore adapter config and WA sandbox number which work with netcore
4) can you please give me access to restart the ODK server instead of requesting you every time?

cc @vinukumar-vs  @ChakshuGautam

**Reply by aayushkedia95** [reply]
Hey @Nik720, 

On 4, please access this URL on any browser to refresh ODK - http://143.110.255.220:10021/odk/updateAll

On the others, can you share your ODK form?

**Reply by Nik720** [reply]
@aayushkedia95 Below points are working now.

> 4. can you please give me access to restart the ODK server instead of requesting you every time?

> 1. with the key-based approach I have to type the same option to get a response and it is working fine, but it is not working with the numeric key as you specified in the documentation that we can still get a response with 1, 2 option

Now key with numeric input is working fine. 
Will I get a response if I type the key which I specified in the name column?

Here is my ODK form https://docs.google.com/spreadsheets/d/1jhlHrRwSdI3mdo8IjoRExIGHuMxj1Wgg2alXYloi-No/edit?usp=sharing
